README.md
# README

1. [koa-passport](https://github.com/rkusa/koa-passport/issues)

[issue fix](https://github.com/rkusa/koa-passport/issues/100#issuecomment-358718370)

2. Authentication done

oauth see [github](https://github.com/ddtramp/oauth-passport-demo) wait passport

https://segmentfault.com/a/1190000011557953

https://www.jianshu.com/p/7010bea0c656

3.  sql inject done

4.  csrf done

5.  koa-compress gzip done http2/push

6.  koa-helmet security

this project use koa-crsf

https://blog.risingstack.com/node-js-security-checklist/

https://segmentfault.com/a/1190000003860400

7.  koa-valivate done parameter

8.  TODO redis cluster

9.  TODO mogodb cluster

10. TODO logic server custer

11. add dev pm2.yml file https://segmentfault.com/q/1010000004131939 done

12. docker dev code

```bash
 docker build -t ddtramdocker/nodemongohttp2:1.0.0 --build-arg mongodb_container_name=mongodb_mongo_1 --build-arg app_env=dev  .

server version:
 docker run --name nodemongohttp2 -p 8000:443 -v /usr/local/Dev/nodejs/nodeMongodDocker/localhost-cert.pem:/usr/src/app/localhost-cert.pem  -v /usr/local/Dev/nodejs/nodeMongodDocker/localhost-privkey.pem:/usr/src/app/localhost-privkey.pem  -v /usr/local/Dev/nodejs/nodeMongodDocker/src:/usr/src/app -v  /usr/local/Dev/nodejs/nodeMongodDocker/package.json:/usr/src/app/package.json -v /usr/local/Dev/nodejs/nodeMongodDocker/node_modules:/usr/src/app/node_modules  --network=mongodb_default -d edfef709a0a6

mount nfs(remote mount files)

  mount -t nfs -o rw,vers=3,tcp,fsc,actimeo=1 192.168.2.132:/Users/jackwang/Documents/Dev /usr/local/Dev/

  -o list , split

sslocal

  docker run -dt --name ss -p 1080:1080 --network=mongodb_default mritd/shadowsocks -m "ss-local" -s "-s 107.182.29.86 -p 8443 -l 1080 -k NWY5ZGMxZj -m aes-256-cfb -b 0.0.0.0 "

docker run -dt --name ssclient -p 1080:1080 mritd/shadowsocks -m "ss-local" -s "-s 127.0.0.1 -p 6500 -b 0.0.0.0 -l 1080 -m chacha20 -k test123 --fast-open" -x -e "kcpclient" -k "-r SSSERVER_IP:6500 -l :6500 -mode fast2"
// todo ss server not done

 local version:
  docker run --name nodemongohttp2 -p 8000:443 -v /Users/jackwang/Documents/Dev/nodejs/nodeMongodDocker/localhost-cert.pem:/usr/src/app/localhost-cert.pem  -v /Users/jackwang/Documents/Dev/nodejs/nodeMongodDocker/localhost-privkey.pem:/usr/src/app/localhost-privkey.pem  -v /Users/jackwang/Documents/Dev/nodejs/nodeMongodDocker/src:/usr/src/app -v  /Users/jackwang/Documents/Dev/nodejs/nodeMongodDocker/package.json:/usr/src/app/package.json -v /Users/jackwang/Documents/Dev/nodejs/nodeMongodDocker/node_modules:/usr/src/app/node_modules  --network=mongodb_default -d a4fd2d9322a2

Enter mongodb

docker exec -ti nodemongo /bin/bash

curl cmd:
curl -X POST -d {a: '123'} localhost:8000/status
curl localhost:8000
curl localhost:8000 -i

curl https://localhost:8000/status -k
curl https://localhost:8000/status -kiv

```

## Maintance

- Logger sys /lib/logger.js

- Health Check /lib/status.js

- Vertical Scaling pm2

- Horizontal Scaling EC2 or AWS

## http2

use `spdy` module

Nodejs build-in module http2 does not work correctly.

Has some issue, it is does not work with this branch.

```bash
rsa key
openssl req -x509 -newkey rsa:2048 -nodes -sha256 -subj '/CN=localhost'  -keyout localhost-privkey.pem -out localhost-cert.pem

```

## Tech

- [http2](https://www.npmjs.com/package/spdy) Nodejs build-in module http2 does not work correctly

- [koa2](https://www.npmjs.com/package/koa2)

- path `build-in`

- [koa-bodyparser](https://www.npmjs.com/package/koa-bodyparser)

- [koa-session2](https://www.npmjs.com/package/koa-session2) session

- [koa-router](https://www.npmjs.com/package/koa-rouuter)

- [koa2-cors](https://www.npmjs.com/package/koa2-cors) Access-Control-Allow-Origin

- [mogodb](https://www.npmjs.com/package/mongo) instrument、 Mongodb native driver

- fs `build-in` file

- [crypto-js](https://www.npmjs.com/package/crypto-js) encrypt string use bcrypt -- deprecated

- [bcrypt](https://www.npmjs.com/package/bcrypt) install error, change to bcryptjs

- [bcryptjs](https://www.npmjs.com/package/bcryptjs)

- [pm2](https://www.npmjs.com/package/pm2) cluster、Operation and maintenance

- [isredis](https://www.npmjs.com/package/ioredis) redis client

- [uid-safe](https://www.npmjs.com/package/uid-safe) uuid

- [pug](https://pugjs.org/api/getting-started.html) html template

- [winston](https://github.com/winstonjs/winston) logger

- [winston-mongodb](https://www.npmjs.com/package/winston-mongodb)

- [winston-daily-rotate-file](https://www.npmjs.com/package/winston-daily-rotate-file)

- [acl](https://github.com/OptimalBits/node_acl) NODE ACL - Access Control Lists for Node

- [koa-crsf](https://www.npmjs.com/package/koa-csrf) Cross site request forgery

- [parameter](https://www.npmjs.com/package/parameter) validate user input

- [jwt|jsonwebtoken](https://www.npmjs.com/package/jsonwebtoken)

- [nodemailer](https://www.npmjs.com/package/nodemailer) send email see lib/send-email.js

- [socks](https://www.npmjs.com/package/socks) for nodemailer proxy, socks support

---

by jack


history.md
# History

## 20200804

- fix koa-passport [issue](https://github.com/rkusa/koa-passport/issues/100#issuecomment-358718370)

- switch http2 -> http; for local test


src/server.js
const http = require("http");
// const spdy = require("spdy");
const Koa2 = require("koa2");
const path = require("path");
const fs = require("fs");

const bodyparser = require("koa-bodyparser");
const Router = require("koa-router");

const rest = require("./lib/rest");
const template = require("./lib/templating");
const Acl = require("./lib/ACL");
const validator = require("./lib/parameter");

// const Http2Puser = require('./lib/http2-pusher') // pusher
const Logger = require("./lib/logger"); // logger
const { logger } = Logger;
const DB = require("./lib/db");

const apiRouter = new Router();
const viewRouter = new Router();

const options = {
  key: fs.readFileSync("./key.pem"),
  cert: fs.readFileSync("./cert.pem"),
};
// Mogodb driver is async
// db must be register before use router
!(async () => {
  // eslint-disable-line
  const db = await DB.initDb();
  const passport = require("./lib/passport-config");
  const controller = require("./lib/controller");

  const app = new Koa2();
  const isProduction = process.env.NODE_ENV === "production";

  const acllogger = {
    // this is acl logger
    debug: (msg) => {
      logger.info(`* ACL DBEUG: ${msg} *`);
    },
  };

  const mongoBackend = new Acl.mongodbBackend(db, "acl_"); // eslint-disable-line
  const acl = new Acl(mongoBackend, acllogger);

  // koa2 error handler, view and api should be separated
  app.use(async (ctx, next) => {
    try {
      await next();
    } catch (err) {
      // TODO record error to logger
      ctx.type = "application/json";
      ctx.status = err.statusCode || err.status || 500;
      ctx.body = {
        message: err.message,
      };
    }
  });

  app.use(async (ctx, next) => {
    ctx.acl = acl;
    ctx.db = db;
    ctx.logger = logger;
    ctx.validator = validator; // only api use this
    await next();
  });

  app.use(async (ctx, next) => {
    ctx.logger.info(`Process ${ctx.request.method} ${ctx.request.url}...`);
    let start = new Date().getTime();
    let execTime;
    await next();
    execTime = new Date().getTime() - start;
    ctx.response.set("X-Response-Time", `${execTime}ms`);
  });

  // static file support:
  if (!isProduction) {
    let staticFiles = require("./lib/static-files");
    app.use(staticFiles("/static/", path.join(__dirname, "/static")));
  }

  app.use(bodyparser());

  // app.use(passport.initialize()); // do not work, see https://github.com/rkusa/koa-passport/issues/100#issuecomment-358718370

  app.use(
    template(path.join(__dirname, "/templates"), {
      globals: [], // https://github.com/pugjs/pug/issues/2141
    })
  );

  app.use(rest.restify());

  controller.addControllers(apiRouter);
  controller.addViewControllers(viewRouter);

  app.use(apiRouter.routes());
  app.use(viewRouter.routes());

  // const server = spdy.createServer(options, app.callback());

  let server = http.createServer(app.callback()).listen(8082, () => {
    // eslint-disable-line
    console.log(`Server is running at 8082`);
  });

  // const boot = () => {
  //   server.listen(443, () => {
  //     // logger.info(`Server is running at ${process.env.PORT}`)
  //     logger.info(`Server is running at ${443}`);
  //   });
  // };

  process.on("uncaughtException", (error) => {
    logger.error("uncaughtException: " + error);
  });

  // if (require.main === module) {
  //   boot();
  // } else {
  //   // for testing
  //   console.log("Running app as a module");
  //   module.exports = {
  //     boot,
  //     shutdown: () => {
  //       server.close(process.exit);
  //     },
  //     port: 443,
  //   };
  // }
})();
